metadata.json

```diff
@@ -30,7 +30,7 @@
     },
     {
       "name": "puppetlabs-stdlib",
-      "version_requirement": ">= 9.0.0 < 10.0.0"
+      "version_requirement": ">= 8.0.0 < 10.0.0"
```

Comment:
Can you please provide more information as to why this threshold has been increased? As per @MartyEwings comment

Reply:
This all relates to behavioural patterns of some customers, so please allow me to break down the events that lead to a certain selection of module versions which leads to a failure mode.

Some customers cannot upgrade to stdlib >= 9,  this is due to breaking changes that were made in this module, mainly around functions that have been renamed/repurposed.  Customers have other modules that need upgrading before they can upgraded to stdlib > 9.  For reasons around capability, resource, attitude to risk etc. they are not capable of doing that at this time.

The customers also adhere to the supported module versions as stated in the dependencies, this is to ensure they have known tested components.  This leads us down the path of the failure mode for this module.

Customers not on stdlib >= 9 have to use puppet_operation_dashboard v2.0.0, as this is the latest listed compatible version.  However, this version no longer works due to a hard coded key for grafana when using Debian - this has since been fixed but only with the supported upgrade to stdlib >= 9.

As it stands, customers who adhere to the versions stated in dependencies that are running stdlib < 9 on Debian systems, the puppet_operational_dashboard module does not work for them.

The change I have implemented may not be the best one, injecting a v2.0.1 with a corrected key may be a better alternative but this would been customers will be locked to a version of puppet_operational_dashboard until they upgrade to stdlib >= 9.

Hope this helps.

Reply:
(I don't work for Puppet)

I think this is the wrong way forward. The users need to update their existing code. The longer they wait the bigger the workload gets but it gets also more complex. And this is a perfect scenario.

> Some customers cannot upgrade to stdlib >= 9, this is due to breaking changes that were made in this module, mainly around functions that have been renamed/repurposed.

I highlighted this in the past. The functions were removed in stdlib 9 and the successors were introduced in the same release. This makes the upgrade more complex. This was noticed after the release but the content team refused to backport the changes to stdlib 8. I still think this was a mistake.

> As it stands, customers who adhere to the versions stated in dependencies that are running stdlib < 9 on Debian systems, the puppet_operational_dashboard module does not work for them.

They brought themself in this situation because they didn't update modules for months, probably years. And now they have an actual problem that needs to be fixed first. Just lowering the stdlib version in metadata.json won't help? That will just tell them it's compatible, but that's a lie. At least the telegraf module has a hard dependency to stdlib 9.

Reply:
In this case, the code in question that is locked on stdlib <=9 was from the forge and and at least one being a puppetlabs module but I recall their being more than one when upgrading to Puppet 8 codebase.  None of their own code has this dependency issue.  I'm trying to get more details on the exact module(s).

> I highlighted this in the past. The functions were removed in stdlib 9 and the successors were introduced in the same release. This makes the upgrade more complex. This was noticed after the release but the content team refused to backport the changes to stdlib 8. I still think this was a mistake.

This is the core issue.  A blocker has been introduced to the upgrade path for stdlib.

> At least the telegraf module has a hard dependency to stdlib 9.

So it does, somehow missed that.

> Creating a new environment for the ops dashboard node to run in with stdlib >= 9.0 and all other dependencies satisfied should also work.

Unfortunately, this goes against their production release policy and was rejected as a workaround.

I had them running their own version of puppet_operational_dashboard with the key correction in place for v2.0.0.